project/apps/web/app/search/page.tsx

```diff
@@ -0,0 +1,237 @@
+      console.log('Match found:', message);
+      setMatchFound(true);
+      stopSearch();
+      toast({
+        variant: 'success',
+        title: 'Match Found',
+        description: 'Your match was successful.',
+      });
+
+      router.push(`/match/${message}`);
+    };
+
+    const handleMatchInvalid = (message: string) => {
+      console.log('Match invalid:', message);
+      stopSearch();
+      stopMatching();
+      toast({
+        variant: 'error',
+        title: 'Match Invalid',
+        description: 'Your match was invalid.',
+      });
+    };
+
+    const handleMatchRequestExpired = (message: string) => {
+      console.log('Match request expired:', message);
+      stopSearch();
+      stopMatching();
+      toast({
+        variant: 'error',
+        title: 'Match Expired',
+        description: 'Your match request has expired. Please try again.',
+      });
+    };
+
+    socket.on('match_found', handleMatchFound);
+    socket.on('match_invalid', handleMatchInvalid);
+    socket.on('match_request_expired', handleMatchRequestExpired);
+
+    return () => {
+      socket.off('match_found', handleMatchFound);
+      socket.off('match_invalid', handleMatchInvalid);
+      socket.off('match_request_expired', handleMatchRequestExpired);
+    };
+  }, [socket, stopSearch, toast, router]);
+
+  // Handle match creation from query params
+  useEffect(() => {
+    if (matchDataParam) {
+      const matchData: MatchRequestMsgDto = JSON.parse(matchDataParam);
+      handleCreateMatch(matchData);
+    }
+  }, [matchDataParam]);
+
+  // Handle browser tab close to cancel match
+  useEffect(() => {
+    const handleBeforeUnload = () => {
+      if (matchReqId) {
+        const matchCancel: MatchCancelDto = { match_req_id: matchReqId };
+        cancelMatchMutation.mutate(matchCancel);
+      }
+    };
+
+    window.addEventListener('beforeunload', handleBeforeUnload);
+
+    return () => {
+      window.removeEventListener('beforeunload', handleBeforeUnload);
+    };
+  }, [matchReqId, cancelMatchMutation]);
+
+  // Connect and disconnect the socket server
+  useEffect(() => {
+    connect();
+
+    return () => {
+      disconnect();
+    };
+  }, [connect, disconnect]);
+
+  // Handle timer
+  useEffect(() => {
+    if (isSearching && intervalRef.current === null) {
+      intervalRef.current = window.setInterval(() => {
+        setTimer((prev) => {
+          return prev + 1;
+        });
+      }, 1000);
+    }
+
+    return () => {
+      if (intervalRef.current !== null) {
+        clearInterval(intervalRef.current);
+        intervalRef.current = null;
+      }
+    };
+  }, [isSearching]);
+
+  // Redirect to home page when isSearching becomes false & no match found
+  useEffect(() => {
+    if (previousIsSearching.current && !isSearching && !matchFound) {
+      router.push('/');
+    }
+
+    previousIsSearching.current = isSearching;
+  }, [isSearching, router, matchFound]);
```

Comment:
from what i see this will only be called upon
1. match cancelled
2. match invalid
3. match expired
should we just add this logic to `handleMatchRequestExpired`, `handleMatchInvalid` and `cancelMatchMutation`'s `onSuccess` instead? prefer to reduce as much `useEffect` as possible due to it having many side effects